File: 2023/Python/01.py
# Advent of Code 2023 - Day 1

def part_one(input):
    total = 0

    # Iterate through each line in text file
    for line in input.split("\n"):
        lineNum = 0

        # Search from left
        for c in line:
            if c.isdigit():
                # insert leftmost digit as tens digit
                lineNum += int(c)*10
                break

        # Search from right
        for c in line[::-1]:
            if c.isdigit():
                # insert rightmost digit as units digit
                lineNum += int(c)
                break

        total += lineNum

    return total

def part_two(input):
    numWords = {"one":1, "two":2, "three":3, "four":4, "five":5, "six":6, "seven":7, "eight":8, "nine":9}
    total = 0

    # Iterate through each line in text file
    for line in input.split("\n"):
        lineNum = 0

        # SEARCH FROM LEFT
        wordFound = False
        # Search from left for first digit char
        for cIndex,c in enumerate(line):
            if wordFound: break

            # First search for num words
            # Search space of up to [1, 2, 3, 4, cIndex]
            for numWordStartPointer in range(max(cIndex - 5, 0), cIndex + 1):
                # Substring not long enough - search deeper into line
                if cIndex - numWordStartPointer + 1 <= 2: break
                if wordFound: break

                # Check each length option
                for wordLen in range(3,6):
                    if line[numWordStartPointer:numWordStartPointer + wordLen] in numWords:
                        # Remember to scale digit by 10 for left side search
                        lineNum += numWords[line[numWordStartPointer:numWordStartPointer + wordLen]] * 10
                        wordFound = True
                        break
            
            # If we find a digit after not finding any words, use digit and break
            if c.isdigit() and not wordFound:
                # insert leftmost digit as tens digit
                lineNum += int(c)*10
                break

        # SEARCH FROM RIGHT
        wordFound = False
        # Search from right
        reverseLine = line[::-1]
        for cIndex,c in enumerate(reverseLine):
            if wordFound: break

            # First search for num words
            # Start the end pointer deepest inside reversed line and move outwards
            for numWordEndPointer in range(cIndex + 1, max(cIndex - 2, 2), -1):
                if wordFound: break

                # Check each length possibility from the end pointer
                for wordLen in range(3,6):
                    # Earliest word can start is at start of reversed line
                    if numWordEndPointer + 1 - wordLen <= 0: break
                    if wordFound: break

                    # Calculating final possible num word string
                    re_reversedSubstring = reverseLine[max(numWordEndPointer - wordLen, 0):numWordEndPointer][::-1]
                    if re_reversedSubstring in numWords:
                        lineNum += numWords[re_reversedSubstring]
                        wordFound = True
                        break

            # If we find a digit after not finding any words, use digit and break
            if c.isdigit() and not wordFound:
                # insert rightmost digit as units digit
                lineNum += int(c)
                break

        # Add 2 digit result of each line to the final total
        total += lineNum

    return total

if __name__ == "__main__":
    with open(f"data/2023/inputs/01.txt", "r") as file:
        input_data = file.read().strip()
    print("Part One:", part_one(input_data))
    print("Part Two:", part_two(input_data))

File: 2023/Rust/01.rs
advent_of_code::solution!(1);

pub fn part_one(input: &str) -> Option<u32> {
    let mut total: u32 = 0;

    for line in input.lines() {
        let mut first_num: u32 = 0;
        let mut last_num: u32 = 0;

        for c in line.chars() {
            if c.is_ascii_digit() {
                // Convert char to u32 using base 10
                last_num = c.to_digit(10).unwrap();
                if first_num == 0 {
                    first_num = c.to_digit(10).unwrap();
                }
            }
        }

        total += (first_num * 10) + last_num;
    }

    Some(total)
}

pub fn part_two(input: &str) -> Option<u32> {
    let mut total: u32 = 0;

    for line in input.lines() {
        let mut new_line = line.replace("one", "one1one");
        new_line = new_line.replace("two", "two2two");
        new_line = new_line.replace("three", "three3three");
        new_line = new_line.replace("four", "four4four");
        new_line = new_line.replace("five", "five5five");
        new_line = new_line.replace("six", "six6six");
        new_line = new_line.replace("seven", "seven7seven");
        new_line = new_line.replace("eight", "eight8eight");
        new_line = new_line.replace("nine", "nine9nine");

        let mut first_num: u32 = 0;
        let mut last_num: u32 = 0;

        for c in new_line.chars() {
            if c.is_ascii_digit() {
                // Convert char to u32 using base 10
                last_num = c.to_digit(10).unwrap();
                if first_num == 0 {
                    first_num = c.to_digit(10).unwrap();
                }
            }
        }

        total += (first_num * 10) + last_num;
    }

    Some(total)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_part_one() {
        let result = part_one(&advent_of_code::template::read_file("examples", DAY));
        assert_eq!(result, None);
    }

    #[test]
    fn test_part_two() {
        let result = part_two(&advent_of_code::template::read_file("examples", DAY));
        assert_eq!(result, None);
    }
}
